src/controller/index.js

```diff
@@ -116,15 +137,186 @@ class Controller extends ENIP {
         this.state.controller.path = PORT.build(BACKPLANE, SLOT);
 
         const sessid = await super.connect(IP_ADDR);
-
         if (!sessid) throw new Error("Failed to Register Session with Controller");
+        
+        this._initializeControllerEventHandlers(); // Connect sendRRData Event
 
-        this._initializeControllerEventHandlers();
+        if (this.state.connectedMessaging === true) {
+            const connid = await this.forwardOpen();
+            if(!connid) throw new Error("Failed to Forward Open with Controller");
+        }
 
         // Fetch Controller Properties and Wall Clock
         await this.readControllerProps();
     }
 
+    /**
+     * Disconnects the PLC instance gracefully by issuing forwardClose, UnregisterSession
+     * and then destroying the socket
+     * and Returns a Promise indicating a success or failure or the disconnection
+     *
+     * @memberof Controller
+     * @returns {Promise}
+     */
+    async disconnect() {
+        if (super.established_conn === true) {
+            const closeid = await this.forwardClose();
+            if(!closeid) throw new Error("Failed to Forward Open with Controller");
+        }
+
+        super.destroy();
+
+        this._removeControllerEventHandlers();
+        return "disconnected";
+    }
+
+    /**
+     * Writes a forwardOpen Request and retrieves the connection ID used for
+     * connected messages.
+     * @memberof Controller
+     * @returns {Promise}
+     */
+    async forwardOpen() {
```

Comment:
Should the `forwardOpen` and `forwardClose` methods be moved up the inheritance chain into the standard `ENIP` class?  Or is this particular implementation specific to Rockwell PLCs?

Reply:
Unfortunately, I am only able to test with RA devices. My trusty point-I/O-Device replies with "Transport class and trigger combination not supported".
Table 3-5.4 of CIP-Volume1 states that forward_open/forward_close are conditional. 
Theoretically, if a device implements the CM-Object and those services, it should work for every device, but I'm not experienced in 3rd-party PLCs / devices.

Reply:
Assuming we move it to ENIP, would we still use the _handleSendRRData/UnitData handlers from the Controller, or would we set up a new set of handlers in the ENIP-class?
This question is related to your comment about the 0x4E Service of Read_Modify_Write_Tag.

Reply:
Okay, I managed to send connected packets to my I/O device, I think we are able to move it to ENIP then, *but* if we do this, we also need to find a way to make the hardware-path (Backplane, Slot) optional, since some CIP-Devices will throw an error if you append it to the packet.

Reply:
Lets hold on moving it for a separate PR.  Though I do think it should be moved. And yes, all event handlers related to those methods would also need to be moved up the inheritance chain as well.  I see what you are saying for the backplane.  I think in the general ENIP class we could receive just a Generic CIP Path.  This way it covers all variants.